Assets/Scripts/SS3D/Systems/Construction/ConstructionMenu.cs

```diff
@@ -0,0 +1,332 @@
+
+        private Vector3 _lastSnappedPosition;
+        private GenericObjectSo _selectedObject;
+
+        private TileSystem _tileSystem;
+        private ConstructionHelper _ghostManager;
+        private Plane _plane;
+
+        private List<GenericObjectSo> _objectDatabase;
+
+        private void Start()
+        {
+            ShowUI(false);
+        }
+
+        private void Initialize()
+        {
+            if (!_initalized)
+            {
+                _tileSystem = SystemLocator.Get<TileSystem>();
+                _ghostManager = GetComponent<ConstructionHelper>();
+                _plane = new Plane(Vector3.up, 0);
+
+                LoadObjectGrid(new[] { TileLayer.Plenum }, false);
+
+                _initalized = true;
+            }
+        }
+
+        private void Update()
+        {
+            // Check for enabling the construction menu
+            if (Input.GetKeyDown(KeyCode.B))
+            {
+                _enabled = !_enabled;
+                ShowUI(_enabled);
+                Initialize();
+            }
+
+            if (!_initalized)
+                return;
+
+            // Clean-up if we are not building
+            if (!_enabled || _selectedObject == null)
+            {
+                _ghostManager.DestroyGhost();
+                return;
+            }
+            else
+            {
+                _ghostManager.CreateGhost(_selectedObject.prefab);
+            }
+
+            if (Input.GetKeyDown(KeyCode.R))
+            {
+                _ghostManager.NextRotation();
+                RefreshGhost();
+            }
+
+
+            // Check if mouse moved
+            Vector3 snappedPosition = TileHelper.GetClosestPosition(GetMousePosition());
+            if (snappedPosition != _lastSnappedPosition && !_itemPlacement)
+            {
+                _ghostManager.SetTargetPosition(snappedPosition);
+                _lastSnappedPosition = snappedPosition;
+                RefreshGhost();
+            }
+
+            if (_itemPlacement)
+            {
+                Vector3 newPosition = GetMousePosition();
+                _ghostManager.SetTargetPosition(newPosition);
+                _lastSnappedPosition = newPosition;
+            }
+
+            // Move ghost
+            _ghostManager.MoveGhost();
+
+            // Check if Left Shift was pressed to replace an existing tile
+            if (Input.GetKeyDown(KeyCode.LeftShift) || Input.GetKeyUp(KeyCode.LeftShift))
+                RefreshGhost();
+
+            // Check if we can want to place or delete an object
+            if (Input.GetKeyDown(KeyCode.Mouse0) && !_mouseOverUI)
+            {
+                if (Input.GetKey(KeyCode.LeftShift))
+                    PlaceObjectClick(_lastSnappedPosition, true);
+                else
+                    PlaceObjectClick(_lastSnappedPosition, false);
+            }
+        }
+
+        private void PlaceObjectClick(Vector3 snappedPosition, bool replaceExisting)
```

Comment:
maybe rename this method, it's confusing that it can delete objects while being called "Place"